packages/lisk-transactions/src/base_transaction.ts

```diff
@@ -0,0 +1,497 @@
+} from './utils';
+import * as schemas from './utils/validation/schema';
+
+const checkTransactionTypes = (
+	tx: TransactionJSON,
+): ReadonlyArray<Error> | undefined => {
+	const typeValidator = validator.compile(schemas.transaction);
+	typeValidator(tx);
+	const transactionErrors = typeValidator.errors
+		? typeValidator.errors.map(
+				error =>
+					new TransactionError(
+						`'${error.dataPath}' ${error.message}`,
+						undefined,
+						error.dataPath,
+					),
+		  )
+		: undefined;
+
+	return transactionErrors;
+};
+
+export interface TransactionResponse {
+	readonly id: string;
+	readonly status: Status;
+	readonly errors: ReadonlyArray<TransactionError>;
+	readonly state?: ReadonlyArray<Account>;
+}
+
+export abstract class BaseTransaction {
+	public readonly amount: BigNum;
+	public readonly fee: BigNum;
+	public readonly id: string;
+	public readonly recipientId: string;
+	public readonly recipientPublicKey: string;
+	public readonly senderId: string;
+	public readonly senderPublicKey: string;
+	public readonly signature?: string;
+	public readonly signatures?: ReadonlyArray<string> = [];
+	public readonly signSignature?: string;
+	public readonly timestamp: number;
+	public readonly type: number;
+	public readonly asset: TransactionAsset = {};
+	public readonly receivedAt: Date = new Date();
+	public readonly isMultiSignature?: boolean = false;
+
+	public constructor(rawTransaction: TransactionJSON) {
+		const result = checkTransactionTypes(rawTransaction);
+		if (result) {
+			throw new TransactionMultiError(
+				'Invalid field types',
+				result as ReadonlyArray<TransactionError>,
+			);
+		}
+
+		this.amount = new BigNum(rawTransaction.amount);
+		this.asset = rawTransaction.asset;
+		this.fee = new BigNum(rawTransaction.fee);
+		this.id = rawTransaction.id;
+		this.recipientId = rawTransaction.recipientId;
+		this.recipientPublicKey = rawTransaction.recipientPublicKey;
+		this.senderId = rawTransaction.senderId;
+		this.senderPublicKey = rawTransaction.senderPublicKey;
+		this.signature = rawTransaction.signature;
+		this.signatures = rawTransaction.signatures;
+		this.signSignature = rawTransaction.signSignature;
+		this.timestamp = rawTransaction.timestamp;
+		this.type = rawTransaction.type;
+		this.receivedAt = rawTransaction.receivedAt;
+		this.isMultiSignature =
+			Array.isArray(rawTransaction.signatures) &&
+			rawTransaction.signatures.length > 0
+				? true
+				: false;
+	}
+
+	public abstract assetToJSON(asset: TransactionAsset): TransactionAsset;
+
+	public toJSON(): TransactionJSON {
+		const transaction = {
+			id: this.id,
+			amount: this.amount.toString(),
+			type: this.type,
+			timestamp: this.timestamp,
+			senderPublicKey: this.senderPublicKey,
+			senderId: this.senderId,
+			recipientId: this.recipientId,
+			recipientPublicKey: this.recipientPublicKey,
+			fee: this.fee.toString(),
+			signatures: this.signatures,
+			asset: this.assetToJSON(this.asset),
+			receivedAt: this.receivedAt,
```

Comment:
```
signature: this.signature ? this.signature : undefined,
signSignature: this.signSignature ? this.signSignature : undefined,
}
```
could work too?

Reply:
would it be better not having `undefined` fields in result JSON?